### PR TITLE
* Add more filters in template functions, so the output of the form on f...

### PR DESCRIPTION
**Pull request by primozcigler**
...rontend can be completely customized (especially tags and classes)

As I mentioned in the email, here is the pull request with newly added filters. This breaks nothing for the existing users, as the output will be the same, but allows to modify all the HTML output if that is necessary (I will find all these new filters useful).

The list of new filters:

* rtb_booking_form_fieldset_tag
* rtb_booking_form_legend_tag
* rtb_booking_form_submit_class
* rtb_booking_form_field_wrapper_class (with 4 different contexts)

### CODE REVIEW

**Comment by NateWr** [comment]
Thanks for the PR @primozcigler. Here's what I'm thinking:

1. Am I right in thinking that you want to filter the classes so that you can add markup classes from a framework, like Bootstrap? I can see that this would be useful if you use this CSS structure. However, I'd prefer a more generic system that allowed you to attach classes to the `$fields` array and have those automatically added. I'm thinking something like:

```
// Modify the $fields array
$fields['fieldset']['callback_args']['classes'] = array( 'my_fieldset_class' );
$fields['fieldset']['fields']['field']['callback_args']['classes'] = array( 'my_field_class' );

// Then it would get passed as the $args to a callback function
if ( !function_exists( 'rtb_print_element_class' ) ) {
function rtb_print_element_class( $slug, $args ) {
	$classes = empty( $args['classes'] ) ? array() : $args['classes'];
	array_push( $classes, $slug );
	return 'class="' . join( ' ', $classes ) .'"';
}
} // endif;
```

2. I have been meaning to print the whole `<button>` element with a callback function. I think I'd prefer to have the whole thing in a callback function and then just let someone overwrite the output with a filter, rather than managing separate filters just for the class. Some people might want an `<input>` tag, for instance. Some might want ajax submission with an `<a>` element. Some might want to add a `<reset>` button.

3. I don't think I want to allow people to modify the `<fieldset>` and `<legend>` element tags. These are not used arbitrarily, but are [important tags for accessibility](http://alistapart.com/article/prettyaccessibleforms#section2). I think if you want to go to the extent of using different wrapper elements, then you need to consider using the `rtb_booking_form_html_pre` filter to output your own form markup.

Are you willing to implement any of the code associated with 1 and 2? If so it'd be great to get an update to your pull request.

**Comment by primozcigler** [comment]
Hi @NateWr,

Yes, I can do this tomorrow, but I will let you know my opinion later on today, I have to run now :)

**Comment by primozcigler** [comment]
Hi again,

So yes, I can implement a and b today or tomorrow - expect the pull request.

While I am aware of the accessibility for the c, I still see no clear reason why would one force the people to use it. If I go with the `rtb_booking_form_html_pre` filter instead I have to write lots of code and throw away all the HTML that was already prepared before. Anyway, I guess 99%+ people will stick with the default markup anyway. It is just another benefit and allows the plugin to be better customized :)

I also don't like the `rtb_booking_form_html_pre` method because if you decide in the future to add/remove fields that we have today, the form might not function properly anymore. Which is not the case if I just alter some wrappers around the main functionality :)

Let me know if you change your mind. Of course you are the author and I will respect your decision, but I think that 2 additional filters cannot do any harm.

**Comment by NateWr** [comment]
> It is just another benefit and allows the plugin to be better customized.

What about if we printed the fieldset with a callback function like the fields? Basically, we just take all of [this](https://github.com/NateWr/restaurant-reservations/blob/master/includes/template-functions.php#L79-L95) and dump it into a callback function that is referenced from the `$fields` array. That way if it's really necessary you can call your own callback function to print the fieldset.

I still want to strongly discourage the abandonment of `fieldset` and `legend`. Is there a styling issue you're having trouble with using the `fieldset` and `legend` tags? Maybe I can help you out there.

> if you decide in the future to add/remove fields that we have today

I take backwards compatibility pretty seriously. The field names which exist are locked in. Should some unforeseen future necessitate a change I'd ensure that there was some kind of backwards compatible system in place. While I would encourage `rtb_booking_form_html_pre` to be treated as a last resort, as long as you're printing the fields in the `$fields` array with the appropriate names and [processing a request](https://github.com/NateWr/restaurant-reservations/blob/master/includes/template-functions.php#L43-L52), you should be able to count on the plugin validating and handling the form properly. I'm even thinking of an addon to use any Ninja Forms or Gravity Form with the plugin (still in conceptual stages).

**Comment by primozcigler** [comment]
Here we go @NateWr. I assume the a. is done just as you thought. I am not so sure for b. so please check it out.

I love to hear you take backwards compatibility seriously. That makes us 2 already :)

I owe you some answers:

> Am I right in thinking that you want to filter the classes so that you can add markup classes from a framework, like Bootstrap?

Bingo, exactly.

> Is there a styling issue you're having trouble with using the fieldset and legend tags? Maybe I can help you out there.

Not really, it won't be a problem anymore now if you merge my PR. It is just that I don't like to target the HTML tags directly in the CSS, I always target classes.

**Comment by NateWr** [comment]
Looks really great, thanks! I've left some comments in the commits for you, just some minor changes here and there.

**Comment by primozcigler** [comment]
Splendid!

I will fix these changes in a new commit in new few minutes.

**Comment by primozcigler** [comment]
Here we go, I think this is it.

**Comment by primozcigler** [comment]
You're right.

File: includes/template-functions.php [hunk 1]
```diff
@@ -345,3 +360,22 @@ function rtb_print_form_error( $field ) {
 	}
 }
 } // endif;
+
+/**
+ * Print a class attribute based on the slug and optional classes, provided with arguments
+ * @since 1.3
+ */
+if ( !function_exists( 'rtb_print_element_class' ) ) {
+function rtb_print_element_class( $slug, $additional_classes ) {
```
Comment (NateWr) [inline includes/template-functions.php]: I'm glad you didn't reverse the order of the arguments. I've changed my mind. Instead, can you just give `$additional_classes` an empty array value?

```
function rtb_print_element_class( $slug, $additional_classes = array() ) {
```

That way if anyone wants to use this function in their own fields they don't have to pass an empty set of classes if they don't want them.

**Comment by NateWr** [comment]
Ok, after those last two things I think we're done. I'm out for the evening but I'll test and merge this tomorrow probably. Nice work!

**Comment by primozcigler** [comment]
I think this is it now.

Any ETA for new release of the plugin?

**Comment by NateWr** [comment]
Thanks @primozcigler! You're the first real code contribution I've gotten so that's exciting.

I am dying to get v1.3 out the door. It's all ready to go (if you wouldn't mind testing the new add/edit bookings modal in the backend that'd be great). However, there is a rather critical issue with the pickadate.js library that needs to get resolved first.

The latest copy of pickadate.js fixes a problem with iOS 8 devices, but the fix introduced even more critical problems for other browsers. I'm working with amsul to try and [get this resolved](https://github.com/amsul/pickadate.js/issues/609) as soon as possible. But as soon as that fix lands I'll be updating the library in the plugin and pushing out v1.3. It's long overdue.

**Comment by primozcigler** [comment]
Yaaay, I am really happy about it and even though you were a bit picky, I totally understand and the final result is much much better than the first proposed solution. And I would do the same in your shoes, so don't take this as criticism, but rather as the right approach ;)

This is the least I can do in return for the awesome work you put in this plugin, so we can use it in our product.

**Pull request merged**